sstables/sstable_set.cc

```diff
@@ -227,21 +232,31 @@ dht::partition_range partitioned_sstable_set::to_partition_range(const dht::ring
 partitioned_sstable_set::partitioned_sstable_set(schema_ptr schema, lw_shared_ptr<sstable_list> all, bool use_level_metadata)
         : _schema(std::move(schema))
         , _all(std::move(all))
-        , _use_level_metadata(use_level_metadata) {
+        , _use_level_metadata(use_level_metadata)
+{
+    _sstable_generations.reserve(_all->size());
+    for (auto& sst : *_all) {
+        auto [it, inserted] = _sstable_generations.emplace(sst->generation());
+        if (!inserted) {
+            on_internal_error(sstlog, format("could not ingest SSTable: detected non-unique sstable generation {}", sst->generation()));
+        }
+    }
```

Comment:
Can you explain why we have to rely on generations? It feels to me like an external facing part of the sstable, ensuring that it has a unique name. I don't understand why the sstable pointer is different from the generation when looking at identity.

Reply:
With this series (at least up to v5 (3ff57fd3d367c345566ae14dc620099b32be50c1))
We'll have 2 shared_sstable instances, with 2 distinct ptrs that essentially point to the same sstable, with the same generation, that are hardlinked to the same inodes.  Therefore just comparing the shared_sstable pointers won't help to establish uniqueness.

Considering your suggestion to clone the staging sstable as an sstable with a new generation - that would indeed require no change of data type (I'll add more to the discussion on the related comment)

Reply:
That said, I still hold my opinion that the sstable generation should be the unique key to a (sub)set of sstables in a table, rather than the shared_sstable pointer, since the design requires sstable generation uniqueness and I don't think we have a clear documented requirement that the shared_sstable instance is unique, i.e. we may open an sstable more than once ending up with more than one shared_sstable instance of the same sstable.

Reply:
Simplified in v6